Make Walkontable Table methods return correct information in the overlay (#6191)

warpech opened the pull request:
### Context
This change is required because my other PR https://github.com/handsontable/handsontable/pull/6157 uses the method `getCell` and expects it to return a TD also in an overlay.

Since `getCell` uses helper methods such as `isRowBeforeRenderedRows`, these methods need to be changed to return correct information in overlays.

The methods that were changed:

```
getFirstRenderedRow
getFirstVisibleRow

getFirstRenderedColumn
getFirstVisibleColumn

getLastRenderedRow
getLastVisibleRow

getLastRenderedColumn
getLastVisibleColumn

isRowBeforeRenderedRows
isRowAfterViewport
isRowAfterRenderedRows

getRenderedColumnsCount
getRenderedRowsCount
getVisibleRowsCount
getVisibleColumnsCount
```

### How has this been tested?

I have added tests for every method changed.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6191
2.
3.

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

jansiegel commented on src/3rdparty/walkontable/src/overlay/table/_base.js:
```diff
@@ -0,0 +1,13 @@
+import Table from '../../table';
+
+class OverlayTable extends Table {
+  alignOverlaysWithTrimmingContainer() {
```
There should probably a comment explaining why these empty methods are here.

warpech replied:
I will solve it differently. These methods are only used in master `Table`, so I will make a class for that and use them there.

jansiegel commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -601,6 +579,7 @@ class Table {
 
   /**
    * Get cell element at coords.
+   * Negative coords.row or coords.cell are used to retrieve header cells
```
`CellCoords` doesn't have a `cell` property, this should probably be `col`.

jansiegel commented on src/3rdparty/walkontable/src/overlay/table/bottom.js:
```diff
@@ -0,0 +1,40 @@
+
+class BottomOverlayTable extends OverlayTable {
+  getFirstRenderedRow() {
+    const index = this.wot.getSetting('totalRows') - this.wot.getSetting('fixedRowsBottom');
```
I think it would be a good idea to add an empty line between the variable declarations and the method logic, for better visual separation. 

This applies to all other methods with this structure.

warpech commented:
Thanks @jansiegel for the review and the improvement suggestions. Below, I will address some doubts about the inconsistencies the way I understand them. 

> - Calling `getCell({row: -1, col: 0})` on any of the overlays throws an `Cannot read property 'childNodes' of undefined` error, while i _think_ it should return the a row header? It does so for the column headers with `col: -1`

I understand now that `getCell` with a negative index should return headers from the current `Table` (overlay or master), if there are any. If there are no headers, and a negative index was given, it will return a negative number error code. I will add it to the JSDoc, implementation and test.

> - When the `fixedRowsTop` option is not defined:
   - `getFirstRenderedRow` returns `0` and `getLastRenderedRow` returns `-1`, which i think is incorrect? I've seen that these methods can return `-1` when no row/column is rendered in the comments in the code, but I think it might be misleading, when `-1` is a valid index (and I think it is, when `getCell` can work on headers?)
    - `getLastRenderedColumn` returns the index of the last rendered column on the `master` overlay, so my understanding is it takes the overlay's headers into calculation (as they're the only thing rendered on the overlay), so maybe `getFirstRendereredRow` should return `-1`, as it is _technically_ the first rendered thing? I'm not sure though.

I understand now that methods 
`getFirstRenderedRow`, `getFirstVisibleRow`, 
`getFirstRenderedColumn`, `getFirstVisibleColumn`, 
`getLastRenderedRow`, `getLastVisibleRow`, 
`getLastRenderedColumn`, `getLastVisibleColumn` should return `-1` if there are no rows or columns in the current `Table` (overlay or master). Headers are irrelevant for these methods. I will add it to the JSDoc, implementation and test.

The methods `isRowBeforeRenderedRows`,
`isRowAfterViewport`, 
`isRowAfterRenderedRows` will always return `false` if there are no rows or columns in the current `Table` (overlay or master).

warpech commented:
> - Most of the methods have the JSDoc comments missing

Thanks, JSDoc added where I think it was missing.

warpech commented:
> - Calling `getCell({row: -1, col: 0})` on any of the overlays throws an `Cannot read property 'childNodes' of undefined` error, while i _think_ it should return the a row header? It does so for the column headers with `col: -1`

Done. Now I have fixed that and the method `getCell` as requested. When given a negative index as the parameter, it returns headers instead of cells. When there is no cell or header found, it returns a negative number exit code.

However, this turned out to be a significant change. Before this PR, the methods `getFirstRenderedRow`, `getFirstVisibleRow`, `getFirstRenderedColumn`, `getFirstVisibleColumn`, `getLastRenderedRow`, `getLastVisibleRow`, `getLastRenderedColumn`, `getLastVisibleColumn` sometimes returned an incorrect value `null`. Some other code in WOT and HOT was treating as a possible answer and used this directly, like here https://github.com/handsontable/handsontable/blob/98878a6bb6edbed065b7961bc97d4b799e908503/src/plugins/autoRowSize/autoRowSize.js#L407

...or indirectly, through type conversion to 0, see here: https://github.com/handsontable/handsontable/blob/98878a6bb6edbed065b7961bc97d4b799e908503/src/3rdparty/walkontable/src/scroll.js#L57-L93

Any untested code that was expecting `undefined` from `getCell or `null` from `getFirstRenderedRow`, `getFirstVisibleRow`, `getFirstRenderedColumn`, `getFirstVisibleColumn`, `getLastRenderedRow`, `getLastVisibleRow`, `getLastRenderedColumn`, `getLastVisibleColumn` might have broken. Please help to verify that.

warpech commented:
> - When the `fixedRowsTop` option is not defined:
>   - `getFirstRenderedRow` returns `0` and `getLastRenderedRow` returns `-1`, which i think is incorrect? I've seen that these methods can return `-1` when no row/column is rendered in the comments in the code, but I think it might be misleading, when `-1` is a valid index (and I think it is, when `getCell` can work on headers?)

According to the docs, the methods `getFirstRenderedRow`, `getFirstVisibleRow`, `getFirstRenderedColumn`, `getFirstVisibleColumn`, `getLastRenderedRow`, `getLastVisibleRow`, `getLastRenderedColumn`, `getLastVisibleColumn` should return an error code `-1` when there are no rows/columns. See this JSDoc here, from the time before this PR: https://github.com/handsontable/handsontable/blob/98878a6bb6edbed065b7961bc97d4b799e908503/src/3rdparty/walkontable/src/table.js#L788-L797. 

I have made more fixes, now when the `fixedRowsTop` option is not defined, both `getFirstRenderedRow` and `getLastRenderedRow` return exit code `-1`. 

I agree that the negative number exit code is unfortunate, because:

- negative indexes can mean headers in other functions
- negative numbers fit well in conditions like `row > this.getLastRenderedRow()`. As a developer, I would expect this comparison to be falsy when `this.getLastRenderedRow()` returns and exit code.

For this, I think that the whole situation with exit codes could be improved by replacing negative number exit codes with ES symbols. Symbols are really cool, because when trying to do `row > this.getLastRenderedRow()` (when the method returns -1) you would simply get and error `TypeError: can't convert symbol to number`.

warpech commented:
> - `getLastRenderedColumn` returns the index of the last rendered column on the `master` overlay, so my understanding is it takes the overlay's headers into calculation (as they're the only thing rendered on the overlay), so maybe `getFirstRendereredRow` should return `-1`, as it is _technically_ the first rendered thing? I'm not sure though.
> - Same thing happens for `leftOverlay` and `topLeftCornerOverlay`

This is fixed now. The methods `getFirstRenderedRow`, `getFirstVisibleRow`, `getFirstRenderedColumn`, `getFirstVisibleColumn`, `getLastRenderedRow`, `getLastVisibleRow`, `getLastRenderedColumn`, `getLastVisibleColumn` return the index on the current overlay, not master.

----

> - For `bottomOverlay`:
>    - `getFirstRenderedRow` returns the index of the last row **plus one**, which is incorrect, as that row does not exist anywhere
>    - `getLastRenderedRow` returns the index of the last rendered row from the `master` overlay, which is also incorrect, as the bottom overlay is empty in this context

This is fixed now.

----

@jansiegel can you please re-review?

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -31,6 +28,7 @@ class Table {
    * @param {HTMLTableElement} table
    */
   constructor(wotInstance, table) {
+    this.isMaster = !wotInstance.cloneOverlay;
```
Could you add jsDoc?

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -635,11 +547,24 @@ class Table {
     }
 
+    if (row < 0) {
+      const zeroBasedHeaderLevel = (-1 * row) - 1;
```
If we set `nestedHeaders` then we get something like that:
![image](https://user-images.githubusercontent.com/11849621/63941099-2dc97680-ca6b-11e9-8242-e3ecd6be27c9.png)

and to get `TH` for header ` N ` we have to call `getCell(-4, 0)` but in my opinion it should be `getCell(-1, 0)`. What do you think?

swistach replied:
BTW. What error code I should get if the cell is in a row before viewport and in column after viewport?
(good catch @wojciechczerniak! :smile::+1:)

warpech replied:
> BTW. What error code I should get if the cell is in a row before viewport and in column after viewport?

The method `getCell` only returns one error code. It validates the input parameters in order in which they are given. Thus, row is validated first and you will get the error code for the row before viewport.

warpech replied:
> and to get TH for header N we have to call getCell(-4, 0) but in my opinion it should be getCell(-1, 0). What do you think?

I guess negative numbers always will be a bit confusing, whatever way we go.

The current order (`-4, 0` returns `N`) comes from the fact, in the Nested Headers concept of "levels", level 0 corresponds to the row that has `A`, level 1 to `D`, level 2 to `H`, and level 3 to `N`.

Thus, to retrieve nested header level `n` from `getCell`, you need to pass `(n + 1) * -1` as the row index.

Of course, I can change that to be `(levelsCount - n) * -1` but are you sure that would be clearer?

For now, I have not changed the behavior of `getCell`, but added a proper test in 541bcb36295e7565dc1d19fa5043c7fa212d90a3.

@swistach, @jansiegel, @wojciechczerniak WDYT?

jansiegel replied:
I think `N` should be obtained with `getCell(-1, 0)` too. The concept of Nested Headers' levels is connected the the fact that they're nested, so higher "level" = nested deeper, but I don't think it translates too well for `getCell` and might potentially be confusing.

warpech replied:
No problem, I have just pushed an implementation that reverses the behavior: https://github.com/handsontable/handsontable/pull/6200/commits/07aef23282374b4a3e630d6ca1968b6f51aa047a

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
   /**
-   * @returns {Number} Returns -1 if no row is visible
+   * 0-based index of column header
+   * @param {Number} level
```
```suggestion
   *
   * @param {Number} level
   * @returns {Boolean}
```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
+  isColumnHeaderLevelRendered(level) {
+    const columnHeaders = this.wot.getSetting('columnHeaders');
+    const columnHeadersCount = columnHeaders.length;
+    return level > (columnHeadersCount - 1);
```
```suggestion

    return level > (columnHeadersCount - 1);
```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
   /**
-   * @returns {Number} Returns -1 if no row is visible, otherwise source index of the last rendered row
+   * 0-based index of row header
+   * @param {Number} level
```
```suggestion
   *
   * @param {Number} level
```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
-    return this.wot.wtViewport.rowsVisibleCalculator.endRow;
-  }
+  isRowBeforeRenderedRows(row) {
+    const first = this.getFirstRenderedRow();
```
```suggestion
    const first = this.getFirstRenderedRow();

```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
+  isRowBeforeRenderedRows(row) {
+    const first = this.getFirstRenderedRow();
+    if (row < 0) {
+      if (first === 0) {
```
You can use `return first !== 0;` instead.

warpech replied:
But then the inline comment that is on the next line will be more difficult to comprehend. Do you think that I remove the comment?

swistach replied:
It shouldn't be `return first <= 0;`? 🤔

I think the comment would be useful when we back here next time.

warpech replied:
Good point. Actually, I think the logic in this function was good but messy. The comment was incorrect, though.

Anyway, I have simplified the logic and explained it in JSDoc: 0d527006fd5e79f2f135e218289722ab03edfdee

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
-      || Overlay.isOverlayTypeOf(this.wot.cloneOverlay, Overlay.CLONE_BOTTOM_LEFT_CORNER)) {
-      return false;
+    if (row < 0) {
+      const zeroBasedHeaderLevel = (-1 * row) - 1;
```
The same remark as for `getCell()`.

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
 
   isColumnBeforeRenderedColumns(column) {
-    return this.columnFilter && (this.columnFilter.sourceColumnToVisibleRowHeadedColumn(column) < 0 && column >= 0);
+    const first = this.getFirstRenderedColumn();
```
```suggestion
    const first = this.getFirstRenderedColumn();

```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
-    return this.columnFilter && (this.columnFilter.sourceColumnToVisibleRowHeadedColumn(column) < 0 && column >= 0);
+    const first = this.getFirstRenderedColumn();
+    if (column < 0) {
+      if (first === 0) {
```
The same remark as for `getLastVisibleRow()`.

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -773,83 +698,87 @@ class Table {
 
   isColumnAfterRenderedColumns(column) {
+    if (column < 0) {
+      const zeroBasedHeaderLevel = (-1 * column) - 1;
```
The same remark as for `getCell()`.

warpech replied:
Just to make sure, you mean this one: https://github.com/handsontable/handsontable/pull/6200#discussion_r319050906?

swistach replied:
Yes, exactly this one 😉

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+const calculatedColumns = {
+  /**
+   * Get the source index of the first rendered column. If no columns are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+ * helper methods that are related to columns.
+ * This mixin is meant to be applied in the subclasses of `Table`
+ * that use virtual rendering in the horizontal axis.
+ * @type {Object}
```
```suggestion
 *
 * @type {Object}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the last column fully visible in the viewport. If no columns are fully visible, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the number of rendered columns.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the number of fully visible columns in the viewport.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+ * helper methods that are related to rows.
+ * This mixin is meant to be applied in the subclasses of `Table`
+ * that use virtual rendering in the vertical axis.
+ * @type {Object}
```
```suggestion
 *
 * @type {Object}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+const calculatedRows = {
+  /**
+   * Get the source index of the first rendered row. If no rows are rendered, returns an error code: -1
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the first row fully visible in the viewport. If no rows are fully visible, returns an error code: -1
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the last rendered row. If no rows are rendered, returns an error code: -1
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the last row fully visible in the viewport. If no rows are fully visible, returns an error code: -1
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the number of rendered rows
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedRows.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the number of fully visible rows in the viewport
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+ * helper methods that are related to columns.
+ * This mixin is meant to be applied in the subclasses of `Table`
+ * that use sticky rendering of the first columns in the horizontal axis.
+ * @type {Object}
```
```suggestion
 *
 * @type {Object}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+const stickyColumnsLeft = {
+  /**
+   * Get the source index of the first rendered column. If no columns are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the source index of the first column fully visible in the viewport. If no columns are fully visible, returns an error code: -1.
+   * Assumes that all rendered columns are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Get the source index of the last rendered column. If no columns are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the source index of the last column fully visible in the viewport. If no columns are fully visible, returns an error code: -1.
+   * Assumes that all rendered columns are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Get the number of rendered columns.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyColumnsLeft.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the number of fully visible columns in the viewport.
+   * Assumes that all rendered columns are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+ * helper methods that are related to rows.
+ * This mixin is meant to be applied in the subclasses of `Table`
+ * that use sticky rendering of the bottom rows in the vertical axis.
+ * @type {Object}
```
```suggestion
 *
 * @type {Object}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+const stickyRowsBottom = {
+  /**
+   * Get the source index of the first rendered row. If no rows are rendered, returns an error code: -1
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Get the source index of the first row fully visible in the viewport. If no rows are fully visible, returns an error code: -1
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+
+  /**
+   * Get the source index of the last rendered row. If no rows are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Get the source index of the last row fully visible in the viewport. If no rows are fully visible, returns an error code: -1.
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+
+  /**
+   * Get the number of rendered rows.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsBottom.js:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Get the number of fully visible rows in the viewport.
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+ * helper methods that are related to rows.
+ * This mixin is meant to be applied in the subclasses of `Table`
+ * that use sticky rendering of the top rows in the vertical axis.
+ * @type {Object}
```
```suggestion
 *
 * @type {Object}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+const stickyRowsTop = {
+  /**
+   * Get the source index of the first rendered row. If no rows are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the source index of the first row fully visible in the viewport. If no rows are fully visible, returns an error code: -1.
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Get the source index of the last rendered row. If no rows are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the source index of the last row fully visible in the viewport. If no rows are fully visible, returns an error code: -1.
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Get the number of rendered rows.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/stickyRowsTop.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Get the number of fully visible rows in the viewport.
+   * Assumes that all rendered rows are fully visible.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/test/spec/table.spec.js:
```diff
@@ -203,10 +203,10 @@ describe('WalkontableTable', () => {
     // expectWtTable(wt, wtTable => wtTable.getCell({row: -1, col: -1}), 'bottom').toBe(undefined); // TODO this should be -1 or not a callable method, but it throws
```
Is it still a TODO thing?

warpech replied:
I think not, I will remove this line.

However, the line 202 is still a TODO.

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -68,7 +66,9 @@ class Table {
 
     this.wtRootElement = this.holder.parentNode;
-    this.alignOverlaysWithTrimmingContainer();
+    if (this.isMaster) {
```
```suggestion

    if (this.isMaster) {
```

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -93,6 +93,16 @@ class Table {
+  /**
+   * Returns a boolean that is true if this intance of Table represents a specific overlay, identified by the overlay name.
+   * For MasterTable, it returns false.
+   * @param {String} overlayTypeName
```
```suggestion
   *
   * @param {String} overlayTypeName
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the last rendered column. If no columns are rendered, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/src/table/mixin/calculatedColumns.js:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Get the source index of the first column fully visible in the viewport. If no columns are fully visible, returns an error code: -1.
+   * @returns {Number}
```
```suggestion
   *
   * @returns {Number}
```

swistach commented on src/3rdparty/walkontable/test/spec/table.spec.js:
```diff
@@ -1213,7 +1586,29 @@ describe('WalkontableTable', () => {
       wt.draw();
 
-      expect(wt.wtTable.getFirstRenderedRow()).toBe(3);
+      expect(wt.wtTable.getLastRenderedRow()).toBe(11); // TODO I think this should be 10, investigate
```
Did you investigate this behaviour?

warpech replied:
No, I think that would be beyond the scope of this change. The line in question was changed from performing a wrong assertion into performing a correct assertion of the current code.

swistach replied:
Alright then. I guess we have to investigate why `ViewportRowsCalculator` returns the incorrect number of rows to render.

swistach commented on src/editorManager.js:
```diff
@@ -142,7 +142,11 @@ class EditorManager {
 
     if (editorClass) {
       this.activeEditor = getEditorInstance(editorClass, this.instance);
+      this.activeEditor.row = row; // pre-preparation needed by getEditedCell
```
Maybe we can use `this.instance.getCell` with `topMost` flag to get `TD` element?

warpech replied:
Done in 9c0f490d2a0b89473ac5576a11725dd71f8499cc

swistach commented on src/editors/textEditor.js:
```diff
@@ -286,7 +286,7 @@ class TextEditor extends BaseEditor {
     // TD is outside of the viewport.
     if (!this.TD) {
       if (!force) {
-        this.close();
```
I guess these issues are related: #3557, #3656 and definitely to do on a separated branch.

swistach commented on test/e2e/editors/textEditor.spec.js:
```diff
@@ -224,7 +224,7 @@ describe('TextEditor', () => {
   });
 
-  it('should hide whole editor when it is higher then header', (done) => {
+  it('should hide whole editor when it is higher then header and TD is not rendered anymore', (done) => {
```
Please use
```
() => async() {
   ...
   await sleep(...);
   ...
}
```
instead of
```
() =>(done) {
   ...
   setTimeout(() => {
      done();
   }, ...);
   ...
}
```

swistach commented on test/e2e/editors/textEditor.spec.js:
```diff
@@ -243,8 +243,33 @@ describe('TextEditor', () => {
+    }, 200);
+  });
+
+  it('should hide whole editor when it is higher then header and TD is still rendered', (done) => {
```
The same remark as for test from line 227.

warpech commented:
> And then in our documentation, it'll be rendered like that: https://handsontable.com/docs/7.1.1/Core.html#addHook

Is Walkontable API rendered in the online documentation at all? It seems to me that it's not.

swistach commented:
It is, partially, but it is, eg.: https://handsontable.com/docs/7.1.1/CellRange.html 
In the new version of the documentation, it will be as a separated module. I guess it'll be after LERNA (@wojciechczerniak ?).

warpech commented:
> Please remember following our JSDoc comments style:

All this manual work for formatting JSDoc comments is very tedious both for the developer and the reviewer. Looks like a job for `eslint --fix`. I think there is a package for that: https://www.npmjs.com/package/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-newline-after-description

---

I have addressed all of the comments related to the coding style.

Still need to be address the following comments tomorrow: 

- [ ] https://github.com/handsontable/handsontable/pull/6200#discussion_r319064155, https://github.com/handsontable/handsontable/pull/6200#discussion_r319065340
- [ ] https://github.com/handsontable/handsontable/pull/6200#discussion_r319050906,  https://github.com/handsontable/handsontable/pull/6200#discussion_r319065673
- [ ] https://github.com/handsontable/handsontable/pull/6200#discussion_r319089333

wojciechczerniak commented:
> It is, partially, but it is, eg.: https://handsontable.com/docs/7.1.1/CellRange.html
> In the new version of the documentation, it will be as a separated module. I guess it'll be after LERNA (@wojciechczerniak ?).

Sooner than that. We have to prepare new docs for the Technical Writer before November.

warpech commented:
Waiting for response in https://github.com/handsontable/handsontable/pull/6200#discussion_r319064155, https://github.com/handsontable/handsontable/pull/6200#discussion_r319050906.

warpech commented:
It seems to me that all comments are resolved, right? @wojciechczerniak, @pnowak, @jansiegel

swistach commented on src/3rdparty/walkontable/src/table.js:
```diff
@@ -93,6 +99,17 @@ class Table {
+   * For MasterTable, it returns false.
+   *
+   * @param {String} overlayTypeName
+   * @return {Boolean}
```
```suggestion
   * @returns {Boolean}
```
